lib/base-compiler.js

```diff
@@ -791,25 +796,27 @@ class BaseCompiler {
         return demangler.process(result);
     }
 
-    processOptOutput(hasOptOutput, optPath) {
-        let output = [];
+    async processOptOutput(optPath) {
```

Comment:
this whole function can be greatly simplified using async iteration on the stream created by `fs.createReadStream`

Reply:
In case you aren't familiar with what I mean here, I implemented it as an example of what it might look like.

```js
    async processOptOutput(optPath) {
        const output = [];

        const optStream = fs.createReadStream(optPath, {encoding: "utf-8"})
            .pipe(new compilerOptInfo.LLVMOptTransformer());

        for await (const opt of optStream) {
            if (opt.DebugLoc && opt.DebugLoc.File && opt.DebugLoc.File.indexOf(this.compileFilename) > -1) {
                output.push(opt);
            }
        }

        if (this.compiler.demangler) {
            const result = JSON.stringify(output, null, 4);
            try {
                const demangleResult = await this.exec(
                    this.compiler.demangler, ["-n", "-p"], {input: result});
                return JSON.parse(demangleResult.stdout);
            } catch (exception) {
                // swallow exception and return non-demangled output
                logger.warn(`Caught exception ${exception} during opt demangle parsing`);
            }
        }

        return output;
    }
```

Its an ES2018 feature, and [has been supported in node.js since 10.3](https://node.green/#ES2018-features-Asynchronous-Iterators)

Reply:
Oh fabulous! will do!

Reply:
This is _so_ neat! Thanks for pointing it out!